odl/solvers/smooth/nonlinear_cg.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2014-2017 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Nonlinear version of the conjugate gradient."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+
+from odl.solvers.util import ConstantLineSearch
+
+
+__all__ = ('conjugate_gradient_nonlinear',)
+
+
+def conjugate_gradient_nonlinear(f, x, line_search=1.0, maxiter=1000, nreset=0,
+                                 tol=1e-16, beta_method='FR',
+                                 callback=None):
+    """Conjugate gradient method for general nonlinear problems.
+
+    Parameters
+    ----------
+    f : `Functional`
+        Functional with ``f.gradient``.
+    x : ``op.domain`` element
+        Vector to which the result is written. Its initial value is
+        used as starting point of the iteration, and its values are
+        updated in each iteration step.
+    line_search : float or `LineSearch`, optional
+        Strategy to choose the step length. If a float is given, it is used as
+        a fixed step length.
+    maxiter : int, optional
+        Maximum number of iterations to perform.
+    nreset : int, optional
+        Number of times the solver should be reset. Default: no reset.
+    tol : float, optional
+        Tolerance that should be used for terminating the iteration.
+    beta_method : {'FR', 'PR', 'HS', 'DY'}
```

Comment:
We do that only if it's not clear from the signature, i.e., when the parameter is in `kwargs`. Or if the default is `None` and we need to explain what happens in that case.

The convention for "limited-range" parameters is to name the default first. So I'd say it's fine as it is.